lib/ansible/galaxy/dependency_resolution/dataclasses.py

```diff
@@ -0,0 +1,401 @@
+            req_version = art_mgr.get_direct_collection_version(tmp_inst_req)
+
+        assert req_name is None or not os.path.isdir(req_name)
+        return cls(
+            req_name, req_version,
+            req_source, req_type,
+        )
+
+
+    def __repr__(self):
+        return (
+            '<{self!s} of type {self.type!r} from {src!s}>'.
+            format(self=self, src=self.src or 'Galaxy')
+        )
+
+
+    def __str__(self):
+        return to_native(self.__unicode__())
+
+
+    def __unicode__(self):
+        if self.fqcn is None:
+            return (
+                u'"virtual collection Git repo"' if self.is_scm
+                else u'"virtual collection namespace"'
+            )
+
+        return (
+            u'{fqcn!s}:{ver!s}'.
+            format(fqcn=to_text(self.fqcn), ver=to_text(self.ver))
+        )
+
+    def _get_separate_ns_n_name(self):  # FIXME: use LRU cache
+        return self.fqcn.split('.')
+
+    @property
+    def namespace(self):
+        if self.is_virtual:
+            raise TypeError('Virtual collections do not have a namespace')
+
+        return self._get_separate_ns_n_name()[0]
+
+    @property
+    def name(self):
+        if self.is_virtual:
+            raise TypeError('Virtual collections do not have a name')
+
+        return self._get_separate_ns_n_name()[-1]
+
+    @property
+    def canonical_package_id(self):
+        if not self.is_virtual:
+            return to_native(self.fqcn)
+
+        return (
+            '<virtual namespace from {src!s} of type {src_type!s}>'.
+            format(src=to_native(self.src), src_type=to_native(self.type))
+        )
+
+    @property
+    def is_virtual(self):
+        return self.is_scm or self.is_subdirs
+
+    @property
+    def is_file(self):
+        # FIXME: Make the checks less dynamic assuming that the type is
+        # FIXME: always set in the initializer (and is not None).
+        return self.type == 'file' or (
+            not self.type and
+            _is_file_path(self.src)
+        )
+
+    @property
+    def is_dir(self):
+        return self.type == 'dir' or (
+            not self.type and
+            _is_collection_dir(self.src)
+        )
+
+    @property
+    def namespace_collection_paths(self):
+        assert self.is_subdirs, 'This property only makes sense for namespaces'
+        return [
+            to_native(path)
+            for path in _find_collections_in_subdirs(self.src)
+        ]
+
+    @property
+    def is_subdirs(self):
+        return self.type == 'subdirs' or (
+            not self.type and
+            _is_collection_namespace_dir(self.src)
+        )
+
+    @property
+    def is_url(self):
+        return self.type == 'url' or (
+            not self.type and
+            _is_http_url(self.src)
+        )
+
+    @property
+    def is_scm(self):
+        return self.type == 'git' or (
+            not self.type and
+            _is_git_url(self.src)
+        )
```

Comment:
FIXME: The `or ...` part should probably be removed. There's no reason for this legacy approach to be kept + we only use custom constructors that always set the correct type anyway.